### ISSUE TITLE
Swagger Importer

### ISSUE THREAD

**Issue by marklawlor** [marklawlor/jaws-swagger#2]
JAWS-Swagger should have the ability to create AWSM modules from a valid Swagger file.

**Comment by chrisbaldauf** [comment]
Hi @marklawlor, I've put together some prototype code for this spanning my own unreleased jaws-swagger project and some proposed feature enhancements to jaws core. I'd be happy to contribute anything that I can to this project.

Right now, my first [PR](https://github.com/jaws-framework/JAWS/pull/244) is in the JAWS backlog, but once that gets merged, I have a follow on to add support for specifying path and query parameters. So if my changes are accepted, you'll be able to call `jaws module create foo bar --method POST --path v1/{foo}/{bar} --query-params page,num`. From there, the swagger parsing and jaws execution code was pretty straightforward.

Let me know if you'd like to work together on this, I'm very interested in seeing swagger support added to JAWS.

**Comment by marklawlor** [comment]
Nice! The biggest contributions this project needs is actually PR like that, as I'd prefer to import/export AWSM modules programmatically though JAWS instead of manually updating the json files in the filesystem.

Unfortunately it looks like development on JAWS's master branch as stalled until they finished their refactor.  Until they start accepting PRs I'm working on the export/swagger-ui functionality which can be largely independent of JAWS (but will have to eventually tie back in).

The project is very poorly documented, as I was more focusing on getting the swagger-ui prototype working, but I plan on fleshing everything out with unit tests / comments / documentation by this weekend.

**Comment by chrisbaldauf** [comment]
Cool. What I can do is fork this project and create a branch for the jaws project templating. I'll also file my PR for JAWS, so at a minimum, someone could download the branch and build it themselves. We can at least get some conversations going to make sure we're generally on the same page.

**Comment by marklawlor** [comment]
Sounds good. I think the import will be fairly straight forward, but I'm not exactly to sure how to handle

- $refs. Do we just [deference the swagger.json file?](https://github.com/BigstickCarpet/swagger-parser/blob/master/docs/swagger-parser.md#dereferenceapi-options-callback). This could create a lot of duplicate code but I don't see another way.
- Handling 'X-' headers. We need to support the same ones as [Amazon did with their swagger importer](https://github.com/awslabs/aws-apigateway-importer), but what do we do if theres others? Strip the 'X-' and attach them to the AWSM.json?

**Comment by chrisbaldauf** [comment]
* Yeah, I punted on $refs for the first version. We'll definitely need to
address that.
* I worry about the duplication and upkeep of trying to stay up-to-date
with the amazon swagger importer. It seems to be fairly duplicative of the
swagger spec - is it honored by the swagger libraries, like
[swagger-codegen](https://github.com/swagger-api/swagger-codegen)? What is
modeled in the x-amazon headers that couldn't be transformed from pure
swagger?

**Comment by marklawlor** [comment]
Mostly stuff in the integration steps, so:
* Mapping Templates (Request and Response)
* Caching
* Some of the Auth stuff

Now that I think about it, we could initially skip this stuff as most people will simply using Swagger to setup their endpoints and do the rest in JAWS.

**Comment by chrisbaldauf** [comment]
I think that makes a lot of sense, at least for an initial version. Then we
can extend where we need to. FWIW, I my code is building the
requestTemplates and requestParameters objects based on the path and
queryParams passed in. Hopefully much of the other properties can be
provided in a similar manner and we can help to hide the gory details and
let customers interact with a standard Swagger doc.
